File: 45-jump-game-ii/45-jump-game-ii.cpp
class Solution {
public:
   int dp[10001];
    int solve(vector<int>&nums,int i,int n){
        if(i==n){
            return dp[i] = 0;
        }
        if(i>n) return 0;
        if(dp[i]!=-1) return dp[i];
        int c = INT_MAX-1;
        for(int j=1;j<=nums[i];j++){
            c = min(c,1+solve(nums,i+j,n));
        }
        return dp[i] = c;
    }
    int jump(vector<int>& nums) {
        memset(dp,-1,sizeof(dp));
        return solve(nums,0,nums.size()-1);
    }
};

File: 45-jump-game-ii/45-jump-game-ii.java
class Solution {
    public int jump(int[] nums) {
        
        return helper(nums, 0, new Integer[nums.length]);
    }
    
    public int helper(int[] nums, int index, Integer[] dp){
        
        if(index>=nums.length-1)
            return 0;
        
        if(dp[index]!=null)
            return dp[index];
        
        int min= Integer.MAX_VALUE-1;
        for(int i=1;i<=nums[index];i++){
            min=Math.min(min, 1+helper(nums, index+i, dp));
        }
        
        dp[index]=min;
        return min;
    }
}